tsl/src/continuous_aggs/common.c

```diff
@@ -263,75 +271,52 @@ process_timebucket_parameters(FuncExpr *fe, ContinuousAggsBucketFunction *bf, bo
 				 errmsg("time bucket function must reference the primary hypertable "
 						"dimension column")));
 
-	if (list_length(fe->args) >= 3)
+	nargs = list_length(fe->args);
+	Assert(nargs >= 2 && nargs <= 5);
+
+	/*
+	 * Process the third argument of the time bucket function. This could be `timezone`, `offset`,
+	 * or `origin`.
+	 *
+	 * Time bucket function variations with 3 and 5 arguments:
+	 *   - time_bucket(width SMALLINT, ts SMALLINT,    offset SMALLINT)
+	 *   - time_bucket(width INTEGER,  ts INTEGER,     offset INTEGER)
+	 *   - time_bucket(width BIGINT,   ts BIGINT,      offset BIGINT)
+	 *   - time_bucket(width INTERVAL, ts DATE,        offset INTERVAL)
+	 *   - time_bucket(width INTERVAL, ts DATE,        origin DATE)
+	 *   - time_bucket(width INTERVAL, ts TIMESTAMPTZ, offset INTERVAL)
+	 *   - time_bucket(width INTERVAL, ts TIMESTAMPTZ, origin TIMESTAMPTZ)
+	 *   - time_bucket(width INTERVAL, ts TIMESTAMPTZ, timezone TEXT, origin TIMESTAMPTZ,
+	 *                 offset INTERVAL)
+	 *   - time_bucket(width INTERVAL, ts TIMESTAMP,   offset INTERVAL)
+	 *   - time_bucket(width INTERVAL, ts TIMESTAMP,   origin TIMESTAMP)
+	 */
+	if (nargs >= 3)
 	{
 		Const *arg = check_time_bucket_argument(lthird(fe->args), "third", process_checks);
-		process_additional_timebucket_parameter(bf, arg);
+		process_additional_timebucket_parameter(bf, arg, &custom_origin);
 	}
 
-	if (list_length(fe->args) >= 4)
+	/*
+	 * Process the fourth and fifth arguments of the time bucket function. This could be `origin` or
+	 * `offset`.
+	 *
+	 * Time bucket function variation with 5 arguments:
+	 *   - time_bucket(width INTERVAL, ts TIMESTAMPTZ, timezone TEXT, origin TIMESTAMPTZ,
+	 *                 offset INTERVAL)
+	 */
+	if (nargs >= 4)
```

Comment:
Why not make it explicit that there can only be 3 or 5 arguments? I'd move this into a single check expecting exactly 5 arguments since 4 arguments should not happen.

Reply:
But FuncExpr arguments is not aligned with the actual function signature, for example this function call: `time_bucket('1 hour', ts, timezone=>'America/Sao_Paulo', offset=>'1 day')` is actually the 5 arguments time bucket but FuncExpr has 4 arguments (FuncExpr->args). The logic to decide what to do with arguments is on `process_additional_timebucket_parameter`.